controller/aks-cluster-config-handler.go

```diff
@@ -498,386 +467,133 @@ func BuildUpstreamClusterState(ctx context.Context, secretsCache wranglerv1.Secr
-	}
 
-	if err = future.WaitForCompletionRef(ctx, clusterClient.Client); err != nil {
-		return c, fmt.Errorf("can't get the AKS cluster create or update future response: %v", err)
-	}
-
-	return future.Result(*clusterClient)
-}
-
-// Check if AKS managed Kubernetes cluster exist
-func existsCluster(ctx context.Context, clusterClient *containersv211.ManagedClustersClient, spec aksv1.AKSClusterConfigSpec) bool {
-	resp, err := clusterClient.Get(ctx, spec.ResourceGroup, spec.ClusterName)
-
-	return err == nil && resp.StatusCode == 200
-}
-
-// Delete AKS managed Kubernetes cluster
-func removeCluster(ctx context.Context, clusterClient *containersv211.ManagedClustersClient, spec aksv1.AKSClusterConfigSpec) (err error) {
-	future, err := clusterClient.Delete(ctx, spec.ResourceGroup, spec.ClusterName)
-	if err != nil {
-		return err
+	if !updateAksCluster {
+		return nil, nil
 	}
 
-	err = future.WaitForCompletionRef(ctx, clusterClient.Client)
+	resourceGroupsClient, err := aks.NewResourceGroupClient(credentials)
 	if err != nil {
-		logrus.Errorf("can't get the AKS cluster create or update future response: %v", err)
-		return err
+		return nil, err
 	}
 
-	logrus.Infof("Cluster %v removed successfully", spec.ClusterName)
-	logrus.Infof("Cluster removal status %v", future.Status())
-
-	return nil
-}
-
-func hasCustomVirtualNetwork(spec aksv1.AKSClusterConfigSpec) bool {
-	return spec.VirtualNetwork != nil && spec.Subnet != nil
-}
-
-func hasAgentPoolProfile(spec aksv1.AKSClusterConfigSpec) bool {
-	return len(spec.NodePools) > 0
-}
-
-func hasLinuxProfile(spec aksv1.AKSClusterConfigSpec) bool {
-	return spec.LinuxAdminUsername != nil && spec.LinuxSSHPublicKey != nil
-}
-
-func getSecrets(secretsCache wranglerv1.SecretCache, secretName string) (string, string, error) {
-	if secretName == "" {
-		return "", "", fmt.Errorf("secret name not provided")
+	if !aks.ExistsResourceGroup(ctx, resourceGroupsClient, spec.ResourceGroup) {
+		logrus.Infof("Resource group [%s] does not exist, creating", spec.ResourceGroup)
+		if err := aks.CreateResourceGroup(ctx, resourceGroupsClient, spec); err != nil {
+			return nil, fmt.Errorf("error during updating resource group %v", err)
+		}
+		logrus.Infof("Resource group [%s] updated successfully", spec.ResourceGroup)
 	}
 
-	ns, id := ParseSecretName(secretName)
-	if ns == "" {
-		ns = "default"
-	}
-	secret, err := secretsCache.Get(ns, id)
+	result, err := aks.CreateOrUpdateCluster(ctx, credentials, resourceClusterClient, spec)
 	if err != nil {
-		return "", "", fmt.Errorf("couldn't find secret [%s] in namespace [%s]", id, ns)
+		return nil, fmt.Errorf("failed to update cluster: %v", err)
 	}
 
-	clientIDBytes := secret.Data["clientId"]
-	clientSecretBytes := secret.Data["clientSecret"]
-	if clientIDBytes == nil || clientSecretBytes == nil {
-		return "", "", fmt.Errorf("invalid secret client data for Azure cloud")
+	if result.ManagedClusterProperties.ProvisioningState == nil ||
```

Comment:
Is `||` right here? `ProvisioningState == nil` means the same thing as `*ProvisioningState == "Succeeded"`?

Reply:
yeap, you are right